Cover web UI being down multiple minutes in worker connection retries

okurz opened the pull request:
Retry connection attempts to web UI for multiple minutes to ensure we
can continue jobs when for example the web UI host reboots.

Related progress issue: https://progress.opensuse.org/issues/57620

kraih commented:
@Martchus Good point. Don't think any 4xx is worth retrying, not even 408 (which is a really weird status that you probably will never see in the wild).

okurz commented:
good idea. I will try that.

Martchus commented:
And remove the dot at the end of the commit title.

okurz commented:
* Fixed dot at EOL of git commit subject message
* Simplified test to better cover individual status codes
* Prevent useless retries on more status codes

What I have prepared but not yet finished: Still retry on 408/425

Martchus commented on lib/OpenQA/Worker/WebUIConnection.pm:
```diff
@@ -314,9 +314,9 @@ sub send {
-            if ($error_code == 404) {
-                # don't retry on 404 errors (in this case we can't expect different
-                # results on further attempts)
+            if (400 <= $error_code && $error_code < 500) {
```
`$error_code >= 400 && $error_code < 500` is much more readable too me (for some reason).

okurz replied:
sorry, not convinced. Only if I must

kraih commented:
Just ignore 408 and 425 imo.

okurz commented:
Ok, well then, do you see anything missing?

Martchus commented on lib/OpenQA/Worker/WebUIConnection.pm:
```diff
@@ -314,9 +314,9 @@ sub send {
-                # don't retry on 404 errors (in this case we can't expect different
-                # results on further attempts)
+            if (400 <= $error_code && $error_code < 500) {
+                # don't retry on most 4xx errors (in this case we can't expect
```
On *most*? It doesn't seem that the code makes any exceptions so far. The test also has a similarly misleading comment (unless I'm missing something). By the way, in my previous review I've mentioned that some possible exceptions but likely we can ignore them for our purposes.

okurz replied:
With the latest change I also handle 408 and 425 so the comment and test fit better. Unless you want me to not handle 408+425 in a special way then please see the alternative PR #2899

Martchus commented on t/24-worker-webui-connection.t:
```diff
@@ -315,35 +321,37 @@ qr/502 response: some timeout \(remaining tries: 2\).*502 response: some timeout
+            403 => 'Forbidden',
+            404 => 'Not found',
+            418 => 'I\'m a teapot',
+            426 => 'Upgrade Required',
```
Instead of 418 and 426 I'd rather test 499. The bounds are likely more interesting than random values in between (even though these might not be official HTTP status codes).

okurz replied:
But I like 418 :)

kraih replied:
Me too, 418 must stay! :shipit:

okurz replied:
so I did not change this.
